openapi/tools/merge_openapi.py

```diff
@@ -6,126 +7,139 @@
+    the security element that references them.
+    """
+    for k in input_dict['components'].keys():
+        input_dict['components'][k] = sortDictionary(input_dict['components'][k])
+    components = sortDictionary(input_dict['components'])
     result = dict()
-    result['openapi'] = indict['openapi']
-    result['info'] = indict['info']
-    result['paths'] = paths
+    result['openapi'] = input_dict['openapi']
+    result['info'] = input_dict['info']
+    result['paths'] = sortDictionary(input_dict['paths'])
     result['components'] = components
-    result['security'] = indict['security']
+    result['security'] = input_dict['security']
     return result
 
+def sortDictionary(input_dict):
+    return dict(sorted(input_dict.items(), key=lambda t: t[0]))
+
+def validate_openapi(filepath):
+    """Run the OpenApi validator on the combined output"""
+    spec_dict = read_from_filename(filepath)[0]
+    validate_spec(spec_dict)
+
 def main():
     parser = argparse.ArgumentParser(description='Merge parts into an openapi document')
     parser.add_argument('--main', help='Main openapi YAML document containing the outline of the desired result')
     parser.add_argument('--apidir', help='Relative path to the directory of YAML documents to merge')
-    # For now, we just grab everything in apidir subdirectories plus main
     parser.add_argument('--outdir', help='Output directory to write the resulting YAML document to')
     argdict = vars(parser.parse_args())
 
+    # Check the outdir early
+    outdir = argdict['outdir']
+    Path(outdir).mkdir(parents=True, exist_ok=True)
+
     # Pull in the main file
-    mainy = YFile(argdict['main'])
+    main_tree = OpenApiTree(argdict['main'])
 
     # Pull in the parts files
     apidir = argdict['apidir']
     yaml_list = glob(apidir + "/*/*.yaml", recursive = False)
-    party = list(map(YFile, yaml_list))
+    part_tree_list = list(map(OpenApiTree, yaml_list))
 
     # Bail if we did not get a clean load
-    if global_had_load_error:
-        print("Exiting due to load error")
-        return 1
-
-    # Merge each section
-    for p in party:
-        merge(mainy, p, ['paths'])
-        merge(mainy, p, ['components', 'schemas'])
-        merge(mainy, p, ['components', 'parameters'])
-        merge(mainy, p, ['components', 'responses'])
+    for tree in [main_tree] + part_tree_list:
+        if tree.load_failed():
+            print
+            return 1
+
+    # Merge each section. We don't worry about the order of the merges or the order of
+    # the files we loaded. It is ordered below.
+    for p in part_tree_list:
+        merge(main_tree, p, ['paths'])
+        merge(main_tree, p, ['components', 'schemas'])
+        merge(main_tree, p, ['components', 'parameters'])
+        merge(main_tree, p, ['components', 'responses'])
 
     # Put the merged parts into a pretty order
-    result = order_merge(mainy.data)
+    result = order_merge(main_tree.data)
```

Comment:
This looks like you're reaching into the guts of the class and rearranging them. Why not have the class be responsible for maintaining its own invariants and keep `data` private?

Reply:
Good call. I moved this and the sort function into the class.